### DISCUSSION TITLE
POS function containing start-position

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#142] [Q&A]
I have some trouble understanding the BRTPANDT instruction. Do I misunderstand it?
my POS assembly contains the following coding to determining the existence of the start-position:

```
       .globals=0
pos() .locals=15 .expose=pos.pos
    load r14,"entry offset"
    say r14
    itos a3
    say a3

    brtpandt havestart,a3,1
    load a3,1
    isub a0,a0,1
havestart:
...
```

I'd expect that brtpandt would jump to havestart, if a3 is set, but it doesn't it always gets into the "set default value" branch.
the inserted printf shows, **op2R->status.all_type_flags** is always zero, although parm3 (a3) is set. 
the added say statement shows clearly the value of a3.

What is the meaning of 
**op2R->status.all_type_flags**?

```
       START_INSTRUCTION(BRTPANDT_ID_REG_INT) CALC_DISPATCH(3);
            DEBUG("TRACE - BRTPANDT_ID_REG_INT 0x%x R%d %d\n",
                  (unsigned int)REG_IDX(1),
                  (int)REG_IDX(2),(int)op3I);
            printf("OP2R Flag %d %d\n",op2R->status.all_type_flags,op3I);
            if (op2R->status.all_type_flags & op3I) {
                next_pc = current_frame->module->binary + REG_IDX(1);
                CALC_DISPATCH_MANUAL;
            }
            DISPATCH;
```

**Comment by adesutherland** [comment]
Can you post the calling rexx? I want to check that the pos signature is
right (optional parameter) cheers 🙂

**Comment by rvjansen** [comment]
Hi Peter,

My short answer would be: it clears the testcases (but I did not look at the arguments and I am  ot at home. The instruction does an AND so it depends on the set register type to branch or not. It admit pasting this in while doing lots of other things, so it also can be in error. Will look into that later today!

René.

**Comment by Peter-Jacob** [comment]
calling rexx parms look ok to me.

countstr REXX
```
/* rexx */
options levelb

/* countstr(needle,haystack) returns number occurrences of needle in haystack */
countstr: procedure = .int
  arg expose needle = .string, expose haystack = .string /* Pass by reference */
  count=0
  offset=0

  Do i=1 to 100
     say "countstr offset" offset
     offset=pos(needle,haystack,offset+1)
     say "offset is "offset
     if offset=0 then return count
     count=count+1
  end
return count

/* pos() Procedure */
pos: procedure = .int
arg expose needle = .string, expose haystack = .string, offset = .int /* Pass by reference */
```
countstr RXAS

```
/*
 * cREXX COMPILER VERSION : cREXX I0134
 * SOURCE                 : countstr
 * BUILT                  : 2021-10-08 20:11:39
 */

.globals=0

countstr() .locals=9 .expose=countstr.countstr
   * Line 6: needle = .string
   * Line 6: haystack = .string
   * Line 7: count=0
   load r1,0
   * Line 8: offset=0
   load r2,0
   * Line 10: {DO}
   * Line 10: i=1
   load r3,1
   * Line 10: to 100
   load r4,100
l53:
   * Line 10: to 100
   igt r0,r3,r4
   brt l24,r0
   * Line 11: say "countstr offset" offset
   itos r2
   sconcat r5,"countstr offset",r2
   say r5
   * Line 12: offset=pos(needle,haystack,offset+1)
   * Line 12: pos(needle,haystack,offset+1)
   load r5,3
   swap r6,a1
   swap r7,a2
   iadd r8,r2,1
   call r2,pos(),r5
   swap a1,r6
   swap a2,r7
   * Line 13: say "offset is "offset
   itos r2
   concat r8,"offset is ",r2
   say r8
   * Line 14: {IF} offset=0
   ieq r8,r2,0
   brf l47,r8
   * Line 14: {THEN}
   * Line 14: return count
   ret r1
l47:
   * Line 15: count=count+1
   iadd r1,r1,1
   * Line 10: {Implicit "BY 1"}
   inc r3
   * Line 10: {DO-END}
   br l53
l24:
   * Line 17: return count
   ret r1

pos() .expose=pos.pos
```

**Reply by rvjansen** [reply]
the testcase is:
```
/* POS */
options levelb
say "Look for POS OK"
/* These from the Rexx book. */
if pos('day','Saturday') \= 6 then say 'failed in test 1 '
if pos('x','abc def ghi') \= 0 then say 'failed in test 2 '
if pos(' ','abc def ghi') \= 4 then say 'failed in test 3 '
if pos(' ','abc def ghi',5) \= 8 then say 'failed in test 4 '
/* These from Mark Hessling. */
if pos('foo','a foo foo b') \= 3 then say 'failed in test 5 '
if pos('foo','a foo foo',3) \= 3 then say 'failed in test 6 '
if pos('foo','a foo foo',4) \= 7 then say 'failed in test 7 '
if pos('foo','a foo foo b',30) \= 0 then say 'failed in test 8 '
if pos('foo','a foo foo b',1) \= 3 then say 'failed in test 9 '
if pos('','a foo foo b') \= 0 then say 'failed in test 10 ' segfaults
if pos('foo','') \= 0 then say 'failed in test 11 ' segfaults
if pos('','') \= 0 then say 'failed in test 12 ' segfaults
if pos('b' , 'a') \= 0 then say 'failed in test 13 '
if pos('b','b') \= 1 then say 'failed in test 14 '
if pos('b','abc') \= 2 then say 'failed in test 15 '
if pos('b','def') \= 0 then say 'failed in test 16 '
if pos('foo','foo foo b') \= 1 then say 'failed in test 17 '
say "POS OK"

/* function prototype */
pos: procedure = .int
arg string1 = .string, string2 = .string, start = 1
```

all the tests of pos() work

**Reply by rvjansen** [reply]
Look at this small testcase:
```
/* POS */
options levelb
say pos('day','Saturdayday')
say pos('day','Saturdayday',7)

/* function prototype */
pos: procedure = .int
arg expose string1 = .string, expose string2 = .string, start = 1

```

this gives 6 and 9

now look at the trace:
```
TRACE - LOAD R1,3
TRACE - LOAD R2,"day"
TRACE - LOAD R3,"Saturdayday"
TRACE - SETTP R4 0 <-- argument is not set, 0
TRACE - CALL R5,pos,R1
TRACE - BRTPANDT_ID_REG_INT 0xe R16 1
TRACE - LOAD R16,1 <-- so we need to load a 1 for a default start
TRACE - ISUB R13,R13,1  <-- and sub the obe
```
now with a specified start argument:
```
TRACE - LOAD R2,3
TRACE - LOAD R3,"day"
TRACE - LOAD R4,"Saturdayday"
TRACE - LOAD R5,7
TRACE - SETTP R5 1. <-- 1: the register is in use
TRACE - CALL R1,pos,R2
TRACE - BRTPANDT_ID_REG_INT 0xe R16 1
TRACE - COPY R1,R15
TRACE - STRLEN R2 R1
TRACE - IEQ R11,R2,0
TRACE - BRT 0x83,R11
TRACE - COPY R3,R16 <r3 becomes A3
TRACE - DEC R3
TRACE - ILT R11,R3,0
```

an rxdas confirms that register 16 is A3 here:
```
pos()           .locals=13 .expose=pos.pos
                brtpandt lb_e,a3 /* aka r16 */,1             * 0x000000:00c6 if op2.typeflag && op3 true then goto op1
                load a3 /* aka r16 */,1                      * 0x000004:0098 Load op1 with op2
                isub a0 /* aka r13 */,a0 /* aka r13 */,1     * 0x000007:0006 Integer Subtract (op1=op2-op3)
                load r12,0                                   * 0x00000b:0098 Load op1 with op2
```

**Reply by rvjansen** [reply]
I use this instruction like I use JCL condition codes: I compute it in my head until I know that it is right, then I invert it and it works!

**Comment by adesutherland** [comment]
Please try

/* pos() Procedure */
pos: procedure = .int
arg expose needle = .string, expose haystack = .string, offset = 1 /*
Pass by reference */

**Reply by rvjansen** [reply]
yes, tried that, no difference in functionality, but probably in performance. I was meaning to check what Rexx does here; if it is by reference, the callee can change it, doesn't it? Because it is us, that is ok jere?

**Comment by adesutherland** [comment]
Ah the pass by reference comment is wrong, I think, but the = 1 makes it an
optional

**Comment by Peter-Jacob** [comment]
the test sequence is in total disarray on my side! I need to organize my compile/assembly and test proposal first!

**Comment by Peter-Jacob** [comment]
I think I found the discrepancy (TS_POS works, TS_COUNTSTR not): 
this is the relevant ts_pos coding:
```
l29:
   * Line 7: {IF} pos(' ','abc def ghi',5) \= 8
   * Line 7: pos(' ','abc def ghi',5)
   load r2,3
   load r3," "
   settp r3,2
   load r4,"abc def ghi"
   settp r4,2
   load r5,5
   settp r5,1
   call r1,pos(),r2
   ine r5,r1,8
   brf l38,r5
   * Line 7: {THEN}
   * Line 7: say 'failed in test 4 '
   say "failed in test 4 "
l38: 
```
there are  `settp`  instructions for all parameters which are missing in ts_countstr (why I cannot answer):
  ```
* Line 11: offset=pos(needle,haystack,offset+1)
   * Line 11: pos(needle,haystack,offset+1)
   load r5,3
   swap r6,a1
   swap r7,a2
   iadd r8,r2,1
   call r2,pos(),r5
   swap a1,r6
   swap a2,r7
```
both rexx work with non-expose

**Reply by Peter-Jacob** [reply]
added them manually (without knowing their meaning ;) )
```
  settp r6,2
   settp r7,2
   settp r8,1
```

then POS works fine!

**Comment by rvjansen** [comment]
Hmm, now you have me confused; I thought pos and its test worked well already?

René.

**Reply by Peter-Jacob** [reply]
2 different things:
- there was a minor glitch which I corrected
- countstr (using POS) doesn't work, the reason the compilation misses the settp instructions - so different story.
And yes, confusion, confusion everywhere, it has driven me nuts.

**Comment by adesutherland** [comment]
Good - I think that is enough to find the bug

In pos(' ','abc def ghi',5)
the typeflags for the arguments should be

- ' ' - 2 its a string constant and not a variable so I don't care if pos() trashes the value of the argument register
- 'abc ...' - 2 as above
- 5 - 1 this means that it is a specified optional variable

All good!

In pos(needle,haystack,offset+1)

We should have typeflags:

- needle = 0 - POS() must not change the value of the register. The compiler should add settp r,0 - need to check, but a compiler bug I think
- haystack = 0 as above
- 'offset+1' = 1 - specified optional parameter - definitely a compiler bug!
 
In sum, POS is correct the but the compiler is not calling it properly.

Good digging Peter!

**Comment by adesutherland** [comment]
This bug might also be causing "symptoms" in other bif tests - I will fix in develop directly (I hope!)

**Comment by rvjansen** [comment]
Ok great, and thanks Peter!

René.

**Comment by adesutherland** [comment]
Peter I am having trouble reproducing the optional parameter error - can you please double check?

```
/* function prototype */
pos: procedure = .int
arg string1 = .string, string2 = .string, start = 1
```

in ts_countstr - have you definitely got the start = 1? (It is right in ts_pos as checked in)

Thanks!

**Reply by Peter-Jacob** [reply]
well, well it's my fault!
I used the following prototype statement in countstr.rexx
```
/* pos() Procedure */
pos: procedure = .int
arg needle = .string, haystack = .string, offset = .int
```
But it should be:
```
/* pos() Procedure */
pos: procedure = .int
arg needle = .string, haystack = .string, offset = 1
```
It seems, if offsset=.int is defined the compiler assumes it's a mandatory parameter and doesn't add the settp rx,1 statement, but the pos function relies on its correct setting to determine if it is set or not available. 
Once the prototype is set correctly it works fine!

**Reply by adesutherland** [reply]
BOOM! :-) solved - and no fault to anyone!

One day we will have a way of defining header files ...

**Reply by Peter-Jacob** [reply]
I opened an issue just to remember, there can be a problem but we can't solve it in the moment (and don't need to).
